### DISCUSSION TITLE
nanobind vs pybind11: winner pybind11, but why?

### DISCUSSION THREAD

**Discussion by cgnx-zpal** [wjakob/nanobind#243] [General]
I needed to compare the performances of nanobind and pybind11 and I was surprised that the pybind11 was about twice faster in my case. I think I miss something. Could you please point to my code where it could be better?

I used the latest tagged 1.4 version from nanobind with Python 3.10 on a linux64.
The code is here:
https://github.com/cgnx-zpal/vectorbind/tree/main

One result with a release flag using `pytest` to measure:
command: `pytest -rP test_performance.py`
and the release build result, where the first number is for pybind11 and the second is the nanobind:
```text
Returns by reference:
  0.43122770002810284
  1.2800408999901265
Returns by value:
  1.0729455000255257
  1.9072670000023209
```

**Comment by cgnx-zpal** [comment]
It turned out that the nanobind uses size optimization by default. So, when the NOMINSIZE flag is added to the `nanobind_add_module` function in CMakeLists.txt, the nanobind also wins here.
(In my case it is: `nanobind_add_module(Vector_Nano_Bind  Vector_Nano_Bind.cpp Vector_Common_nb.cpp NOMINSIZE)`)

**Reply by wjakob** [reply]
Yeah -- the default assumption is that the binding code is just the wrapper layer, and expensive compilation is done in another library being bound. I'll add some more documentation about this since several users have now run into this issue.

**Reply by wjakob** [reply]
-> https://github.com/wjakob/nanobind/commit/f9bc4dd4e1c1ada17577f5fc6c9e4e454c0f66b8
